zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java

```diff
@@ -684,7 +684,9 @@ public String getMaxPrefixWithQuota(String path) {
     public void addWatch(String basePath, Watcher watcher, int mode) {
         WatcherMode watcherMode = WatcherMode.fromZooDef(mode);
         dataWatches.addWatch(basePath, watcher, watcherMode);
-        childWatches.addWatch(basePath, watcher, watcherMode);
+        if (watcherMode != WatcherMode.PERSISTENT_RECURSIVE) {
+            childWatches.addWatch(basePath, watcher, watcherMode);
+        }
```

Comment:
IIRC we filter out child watch events for PRESISTENT_RECURSIVE in https://github.com/apache/zookeeper/pull/1820.

Do changes in this file optimize the server-side logic while it's not user-visible and has nothing to do with "Support different watch modes on same path"?

Reply:
No. For sole `WatcherMode.PERSISTENT_RECURSIVE`, there is no `EventType.NodeChildrenChanged` delivered as old code do filter in `WatcherManager#triggerWatch`.

https://github.com/apache/zookeeper/blob/de8768807fc7a3fcabc9762b033cf59e695cf14b/zookeeper-server/src/main/java/org/apache/zookeeper/server/watch/WatchManager.java#L137-L141

The above changes do such filter in first place to reduce coding burden in `triggerWatch` phase.